In [162]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [163]:
data = pd.read_csv('H1b_EDA_data.csv',encoding = 'ISO-8859-1')

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (12) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [164]:
data.head()

CASE_NUMBER CASE_STATUS CASE_SUBMITTED DECISION_DATE VISA_CLASS  \
0  I-200-09121-701936   WITHDRAWN       2/5/2015      2/5/2015       H-1B   
1  I-200-09146-796321   CERTIFIED     12/12/2014    12/18/2014       H-1B   
2  I-200-09180-329758   WITHDRAWN       1/3/2013    12/15/2014       H-1B   
3  I-200-09183-259985   CERTIFIED      3/10/2015     3/16/2015       H-1B   
4  I-200-09189-800933   CERTIFIED     12/17/2014    12/23/2014       H-1B   

  EMPLOYMENT_START_DATE EMPLOYMENT_END_DATE                    EMPLOYER_NAME  \
0            02/09/2015          02/28/2015                  MEDTRONIC, INC.   
1            01/05/2015          01/04/2018           UNIVERSITY OF OKLAHOMA   
2            01/07/2013          01/06/2016                       NOKIA INC.   
3            09/07/2015          09/07/2018  OMRON OILFIELD AND MARINE, INC.   
4            06/06/2015          06/05/2018   FEDERAL HOME LOAN MORTGAGE CO.   

  EMPLOYER_CITY EMP_STATE_abb  ... WORKSITE_POSTAL_CODE  YEAR NAIC_CODE  \
0   MINNEAPOLIS            MN  ...                55412  2015    3345.0   
1        NORMAN            OK  ...                73019  2015    6113.0   
2        IRVING            TX  ...                92127  2015    5172.0   
3       HOUSTON            TX  ...                77088  2015    3353.0   
4        MCLEAN            VA  ...                22102  2015    5222.0   

  WAGE_RATE_OF_PAY total_wage Sector_data EMP_STATE_full  \
0          20000 -    19000.0        33.0      Minnesota   
1       85000.00 -    42860.0        61.0       Oklahoma   
2          94000 -    73965.0        51.0          Texas   
3    66000 - 70000    65998.0        33.0          Texas   
4   97000 - 143300    96907.0        52.0       Virginia   

       EMP_State_and_city Worksite_STATE_full   Worksite_State_and_city  
0  MINNEAPOLIS, Minnesota          California  EDEN PRAIRIE, California  
1        NORMAN, Oklahoma            Oklahoma          NORMAN, Oklahoma  
2           IRVING, Texas          California     SAN DIEGO, California  
3          HOUSTON, Texas               Texas            HOUSTON, Texas  
4        MCLEAN, Virginia            Virginia          MCLEAN, Virginia  

[5 rows x 36 columns]

In [165]:
def count_na(column):
    return column.isna().sum()
data.apply(count_na)

CASE_NUMBER                     0
CASE_STATUS                     0
CASE_SUBMITTED                  1
DECISION_DATE                   0
VISA_CLASS                      0
EMPLOYMENT_START_DATE          38
EMPLOYMENT_END_DATE            47
EMPLOYER_NAME                  67
EMPLOYER_CITY                  29
EMP_STATE_abb                   0
EMPLOYER_POSTAL_CODE           57
EMPLOYER_COUNTRY                0
EMPLOYER_PHONE                  1
AGENT_ATTORNEY_NAME        227034
AGENT_ATTORNEY_CITY        840452
AGENT_ATTORNEY_STATE       899957
JOB_TITLE                      11
SOC_CODE                       18
SOC_NAME                       20
FULL_TIME_POSITION              7
PREVAILING_WAGE                 0
PW_UNIT_OF_PAY                  0
WAGE_UNIT_OF_PAY               19
WORKSITE_CITY                  41
WORKSITE_COUNTY              4616
Worksite_STATE_abb             16
WORKSITE_POSTAL_CODE           70
YEAR                            0
NAIC_CODE                       2
WAGE_RATE_OF_P

In [166]:
data.describe()

PREVAILING_WAGE          YEAR     NAIC_CODE    total_wage   Sector_data
count     2.448729e+06  2.448729e+06  2.448727e+06  2.448729e+06  2.448727e+06
mean      7.257600e+04  2.016497e+03  5.263288e+03  7.844724e+04  5.242158e+01
std       1.332444e+05  1.135611e+00  7.477223e+02  5.217561e+05  7.558952e+00
min       0.000000e+00  2.015000e+03  0.000000e+00  0.000000e+00  0.000000e+00
25%       5.674200e+04  2.015000e+03  5.413000e+03  5.871800e+04  5.400000e+01
50%       7.013800e+04  2.016000e+03  5.415000e+03  7.132300e+04  5.400000e+01
75%       8.881600e+04  2.018000e+03  5.415000e+03  8.962700e+04  5.400000e+01
max       2.016227e+08  2.018000e+03  9.999000e+03  2.729626e+08  9.900000e+01

## Preprocessing

1. Delete the unused features. 
2. Calculate the duration of datetime. 
3. Delete the rows with NA. 
4. Do specific preprocessing listed in the following table. 


|Name|NACount|Note|
| :---- | :---- | :---- |
|Sector_data|2|<font color = 'red'>Delete </font>|
|EMP_STATE_full|2358|<font color = 'red'>Delete </font>|
|EMP_State_and_city|0|<font color = 'red'>Delete </font>|
|Worksite_STATE_full|2453|<font color = 'red'>Delete </font>|
|Worksite_State_and_city|0|<font color = 'red'>Delete </font>|
|EMPLOYER_PHONE|1|<font color = 'red'>Delete </font>|
|AGENT_ATTORNEY_NAME|227034|<font color = 'red'>Delete </font>|
|AGENT_ATTORNEY_CITY|840452|<font color = 'red'>Delete </font>|
|AGENT_ATTORNEY_STATE|899957|<font color = 'red'>Delete </font>|
|JOB_TITLE|11|<font color = 'red'>Delete </font>|
|EMPLOYER_POSTAL_CODE|57|<font color = 'red'>Delete </font>|
|WORKSITE_POSTAL_CODE|70|<font color = 'red'>Delete </font>|
|YEAR|0|<font color = 'red'>Delete </font>|
|WORKSITE_COUNTY|4616|<font color = 'red'>Delete </font>|
|WAGE_RATE_OF_PAY|0|<font color = 'red'>Delete </font>|
|CASE_STATUS|0|<font color = 'green'>label </font>|
|CASE_SUBMITTED|1| <font color = 'blue'> calculate duration </font>|
|DECISION_DATE|0|<font color = 'blue'> calculate duration </font>|
|EMPLOYMENT_START_DATE|38|<font color = 'blue'> calculate duration </font>|
|EMPLOYMENT_END_DATE|47|<font color = 'blue'> calculate duration </font>|
|EMPLOYER_COUNTRY|0|<font color = 'blue'> Only consider American. </font>|
|PREVAILING_WAGE|0|<font color = 'blue'> Normalize it to yearly wage. </font>|
|PW_UNIT_OF_PAY|0|<font color = 'blue'> Normalize it to yearly wage. </font>|
|WAGE_UNIT_OF_PAY|19|<font color = 'blue'> Normalize it to yearly wage. </font>|
|total_wage|0|<font color = 'blue'> Normalize it to yearly wage. </font>|
|NAIC_CODE|2|<font color = 'blue'> Keep three digits. </font>|
|EMPLOYER_CITY|29|<font color = 'blue'> For geographical figures.  </font>|
|EMP_STATE_abb|0|<font color = 'blue'> For geographical figures. </font>|
|CASE_NUMBER|0||
|VISA_CLASS|0||
|EMPLOYER_NAME|67||
|SOC_CODE|18||
|SOC_NAME|20||
|FULL_TIME_POSITION|7||
|WORKSITE_CITY|41||
|Worksite_STATE_abb|16||


In [167]:
label = ['CASE_STATUS']
target = data.loc[:, label]

features = ['CASE_SUBMITTED', 'DECISION_DATE', 'EMPLOYMENT_START_DATE', 
            'EMPLOYMENT_END_DATE', 'EMPLOYER_COUNTRY', 'PREVAILING_WAGE', 
           'PW_UNIT_OF_PAY', 'WAGE_UNIT_OF_PAY', 'NAIC_CODE',
           'EMPLOYER_CITY', 'EMP_STATE_abb', 'CASE_NUMBER', 'VISA_CLASS',
           'EMPLOYER_NAME', 'SOC_CODE', 'SOC_NAME', 'FULL_TIME_POSITION',
           'WORKSITE_CITY', 'Worksite_STATE_abb', 'total_wage']
train_vals = data.loc[:, features]

In [168]:
# Drop NA
train_vals = train_vals.dropna()
# Keep rows with employer country = usa
train_vals = train_vals[train_vals['EMPLOYER_COUNTRY'] == 'UNITED STATES OF AMERICA']

In [169]:
def naic_process(code):
    if(code > 100000):
        res = code // 1000
    elif(code > 10000):
        res = code // 100
    elif(code > 1000):
        res = code // 10
    else:
        res = code
    
    return str(int(code))

# Drop rows with NAIC_CODE less than 3 digits
train_vals = train_vals[train_vals['NAIC_CODE'] > 100]
# Keep the first 3 digits of the remaining NAIC_CODE
train_vals.loc[:, 'NAIC_CODE'] = train_vals.loc[:, 'NAIC_CODE'].apply(lambda x: naic_process(x))

In [170]:
import time 
import datetime
def str_to_datetime(string):
    try:
        timestamp = time.strptime(string, '%m/%d/%Y')
        return datetime.datetime(*timestamp[:6])
    except:
        try:
            timestamp = time.strptime(string, '%d/%m/%Y')
            return datetime.datetime(*timestamp[:6])
        except:
            return 

In [171]:
date_label = ['CASE_SUBMITTED', 
              'DECISION_DATE', 
              'EMPLOYMENT_START_DATE', 
              'EMPLOYMENT_END_DATE']
data_date = train_vals.loc[:,date_label]

In [172]:
data_date = data_date.applymap(str_to_datetime)

In [173]:
data_date.head()

CASE_SUBMITTED DECISION_DATE EMPLOYMENT_START_DATE EMPLOYMENT_END_DATE
0     2015-02-05    2015-02-05            2015-02-09          2015-02-28
1     2014-12-12    2014-12-18            2015-01-05          2018-01-04
2     2013-01-03    2014-12-15            2013-01-07          2016-01-06
3     2015-03-10    2015-03-16            2015-09-07          2018-09-07
4     2014-12-17    2014-12-23            2015-06-06          2018-06-05

In [174]:
tmp_column = set(train_vals.columns)
tmp_column = tmp_column - set(date_label)

In [175]:
train_vals = pd.concat([train_vals.loc[:,tmp_column], data_date], axis = 1)

In [176]:
train_vals.head()

EMPLOYER_CITY FULL_TIME_POSITION VISA_CLASS          EMPLOYER_COUNTRY  \
0   MINNEAPOLIS                  Y       H-1B  UNITED STATES OF AMERICA   
1        NORMAN                  Y       H-1B  UNITED STATES OF AMERICA   
2        IRVING                  Y       H-1B  UNITED STATES OF AMERICA   
3       HOUSTON                  Y       H-1B  UNITED STATES OF AMERICA   
4        MCLEAN                  Y       H-1B  UNITED STATES OF AMERICA   

  WAGE_UNIT_OF_PAY EMP_STATE_abb WORKSITE_CITY         CASE_NUMBER  \
0             Year            MN  EDEN PRAIRIE  I-200-09121-701936   
1             Year            OK        NORMAN  I-200-09146-796321   
2             Year            TX     SAN DIEGO  I-200-09180-329758   
3             Year            TX       HOUSTON  I-200-09183-259985   
4             Year            VA        MCLEAN  I-200-09189-800933   

   total_wage                                SOC_NAME Worksite_STATE_abb  \
0     19000.0                    MECHANICAL ENGINEERS                 CA   
1     42860.0     ENGINEERING TEACHERS, POSTSECONDARY                 OK   
2     73965.0        COMPUTER OCCUPATIONS, ALL OTHER*                 CA   
3     65998.0  ELECTRONICS ENGINEERS, EXCEPT COMPUTER                 TX   
4     96907.0                    COMPUTER PROGRAMMERS                 VA   

                     EMPLOYER_NAME  PREVAILING_WAGE    SOC_CODE  \
0                  MEDTRONIC, INC.          19000.0  17-2141.00   
1           UNIVERSITY OF OKLAHOMA          42860.0     25-1032   
2                       NOKIA INC.          73965.0     15-1799   
3  OMRON OILFIELD AND MARINE, INC.          65998.0     17-2072   
4   FEDERAL HOME LOAN MORTGAGE CO.          96907.0     15-1131   

  PW_UNIT_OF_PAY NAIC_CODE CASE_SUBMITTED DECISION_DATE EMPLOYMENT_START_DATE  \
0           Year      3345     2015-02-05    2015-02-05            2015-02-09   
1           Year      6113     2014-12-12    2014-12-18            2015-01-05   
2           Year      5172     2013-01-03    2014-12-15            2013-01-07   
3           Year      3353     2015-03-10    2015-03-16            2015-09-07   
4           Year      5222     2014-12-17    2014-12-23            2015-06-06   

  EMPLOYMENT_END_DATE  
0          2015-02-28  
1          2018-01-04  
2          2016-01-06  
3          2018-09-07  
4          2018-06-05

In [177]:
train_vals['CASE_DURATION'] = train_vals.loc[:,'DECISION_DATE'] - train_vals.loc[:,'CASE_SUBMITTED']
train_vals['EMPLOYMENT_DURATION'] = train_vals.loc[:,'EMPLOYMENT_END_DATE'] - train_vals.loc[:, 'EMPLOYMENT_START_DATE']

In [178]:
def to_days(e):
    return e.days

In [179]:
train_vals.loc[:,'CASE_DURATION'] = train_vals.loc[:,'CASE_DURATION'].apply(to_days) 

In [180]:
train_vals.loc[:,'EMPLOYMENT_DURATION'] = train_vals.loc[:,'EMPLOYMENT_DURATION'].apply(to_days)

In [181]:
train_vals.head()

EMPLOYER_CITY FULL_TIME_POSITION VISA_CLASS          EMPLOYER_COUNTRY  \
0   MINNEAPOLIS                  Y       H-1B  UNITED STATES OF AMERICA   
1        NORMAN                  Y       H-1B  UNITED STATES OF AMERICA   
2        IRVING                  Y       H-1B  UNITED STATES OF AMERICA   
3       HOUSTON                  Y       H-1B  UNITED STATES OF AMERICA   
4        MCLEAN                  Y       H-1B  UNITED STATES OF AMERICA   

  WAGE_UNIT_OF_PAY EMP_STATE_abb WORKSITE_CITY         CASE_NUMBER  \
0             Year            MN  EDEN PRAIRIE  I-200-09121-701936   
1             Year            OK        NORMAN  I-200-09146-796321   
2             Year            TX     SAN DIEGO  I-200-09180-329758   
3             Year            TX       HOUSTON  I-200-09183-259985   
4             Year            VA        MCLEAN  I-200-09189-800933   

   total_wage                                SOC_NAME  ... PREVAILING_WAGE  \
0     19000.0                    MECHANICAL ENGINEERS  ...         19000.0   
1     42860.0     ENGINEERING TEACHERS, POSTSECONDARY  ...         42860.0   
2     73965.0        COMPUTER OCCUPATIONS, ALL OTHER*  ...         73965.0   
3     65998.0  ELECTRONICS ENGINEERS, EXCEPT COMPUTER  ...         65998.0   
4     96907.0                    COMPUTER PROGRAMMERS  ...         96907.0   

     SOC_CODE  PW_UNIT_OF_PAY NAIC_CODE CASE_SUBMITTED DECISION_DATE  \
0  17-2141.00            Year      3345     2015-02-05    2015-02-05   
1     25-1032            Year      6113     2014-12-12    2014-12-18   
2     15-1799            Year      5172     2013-01-03    2014-12-15   
3     17-2072            Year      3353     2015-03-10    2015-03-16   
4     15-1131            Year      5222     2014-12-17    2014-12-23   

  EMPLOYMENT_START_DATE EMPLOYMENT_END_DATE CASE_DURATION EMPLOYMENT_DURATION  
0            2015-02-09          2015-02-28             0                19.0  
1            2015-01-05          2018-01-04             6              1095.0  
2            2013-01-07          2016-01-06           711              1094.0  
3            2015-09-07          2018-09-07             6              1096.0  
4            2015-06-06          2018-06-05             6              1095.0  

[5 rows x 22 columns]

In [182]:
train_vals = train_vals.dropna()

In [183]:
train_vals = train_vals.drop(columns = 'EMPLOYER_COUNTRY')

In [184]:
train_vals = train_vals.drop(columns = date_label)

In [185]:
train_vals.head()

EMPLOYER_CITY FULL_TIME_POSITION VISA_CLASS WAGE_UNIT_OF_PAY EMP_STATE_abb  \
0   MINNEAPOLIS                  Y       H-1B             Year            MN   
1        NORMAN                  Y       H-1B             Year            OK   
2        IRVING                  Y       H-1B             Year            TX   
3       HOUSTON                  Y       H-1B             Year            TX   
4        MCLEAN                  Y       H-1B             Year            VA   

  WORKSITE_CITY         CASE_NUMBER  total_wage  \
0  EDEN PRAIRIE  I-200-09121-701936     19000.0   
1        NORMAN  I-200-09146-796321     42860.0   
2     SAN DIEGO  I-200-09180-329758     73965.0   
3       HOUSTON  I-200-09183-259985     65998.0   
4        MCLEAN  I-200-09189-800933     96907.0   

                                 SOC_NAME Worksite_STATE_abb  \
0                    MECHANICAL ENGINEERS                 CA   
1     ENGINEERING TEACHERS, POSTSECONDARY                 OK   
2        COMPUTER OCCUPATIONS, ALL OTHER*                 CA   
3  ELECTRONICS ENGINEERS, EXCEPT COMPUTER                 TX   
4                    COMPUTER PROGRAMMERS                 VA   

                     EMPLOYER_NAME  PREVAILING_WAGE    SOC_CODE  \
0                  MEDTRONIC, INC.          19000.0  17-2141.00   
1           UNIVERSITY OF OKLAHOMA          42860.0     25-1032   
2                       NOKIA INC.          73965.0     15-1799   
3  OMRON OILFIELD AND MARINE, INC.          65998.0     17-2072   
4   FEDERAL HOME LOAN MORTGAGE CO.          96907.0     15-1131   

  PW_UNIT_OF_PAY NAIC_CODE  CASE_DURATION  EMPLOYMENT_DURATION  
0           Year      3345              0                 19.0  
1           Year      6113              6               1095.0  
2           Year      5172            711               1094.0  
3           Year      3353              6               1096.0  
4           Year      5222              6               1095.0

In [186]:
train_vals.loc[:,'FULL_TIME_POSITION'] = (train_vals.loc[:,'FULL_TIME_POSITION'] == 'Y')

In [187]:
print(np.unique(train_vals['WAGE_UNIT_OF_PAY']))
print(np.unique(train_vals['PW_UNIT_OF_PAY']))

['Bi-Weekly' 'Hour' 'Month' 'Week' 'Year']
['Bi-Weekly' 'Hour' 'Month' 'Week' 'Year']


In [188]:
tmp_multiplier = {
    'Bi-Weekly': 25,
    'Hour': 50*5*8,
    'Month' : 12,
    'Week': 50,
    'Year': 1
}
def to_multiplier(e):
    return tmp_multiplier[e]
train_vals.loc[:,'PW_UNIT_OF_PAY'] = train_vals.loc[:,'PW_UNIT_OF_PAY'].apply(to_multiplier)
train_vals.loc[:,'WAGE_UNIT_OF_PAY'] = train_vals.loc[:,'WAGE_UNIT_OF_PAY'].apply(to_multiplier)
train_vals.loc[:, 'PREVAILING_WAGE'] = train_vals.loc[:, 'PREVAILING_WAGE'] *train_vals.loc[:,'PW_UNIT_OF_PAY']
train_vals.loc[:, 'total_wage'] = train_vals.loc[:, 'total_wage']* train_vals.loc[:,'WAGE_UNIT_OF_PAY']

In [190]:
train_vals = train_vals.drop(columns = ['PW_UNIT_OF_PAY','WAGE_UNIT_OF_PAY'])

In [191]:
train_vals.head()

EMPLOYER_CITY  FULL_TIME_POSITION VISA_CLASS EMP_STATE_abb WORKSITE_CITY  \
0   MINNEAPOLIS                True       H-1B            MN  EDEN PRAIRIE   
1        NORMAN                True       H-1B            OK        NORMAN   
2        IRVING                True       H-1B            TX     SAN DIEGO   
3       HOUSTON                True       H-1B            TX       HOUSTON   
4        MCLEAN                True       H-1B            VA        MCLEAN   

          CASE_NUMBER  total_wage                                SOC_NAME  \
0  I-200-09121-701936     19000.0                    MECHANICAL ENGINEERS   
1  I-200-09146-796321     42860.0     ENGINEERING TEACHERS, POSTSECONDARY   
2  I-200-09180-329758     73965.0        COMPUTER OCCUPATIONS, ALL OTHER*   
3  I-200-09183-259985     65998.0  ELECTRONICS ENGINEERS, EXCEPT COMPUTER   
4  I-200-09189-800933     96907.0                    COMPUTER PROGRAMMERS   

  Worksite_STATE_abb                    EMPLOYER_NAME  PREVAILING_WAGE  \
0                 CA                  MEDTRONIC, INC.          19000.0   
1                 OK           UNIVERSITY OF OKLAHOMA          42860.0   
2                 CA                       NOKIA INC.          73965.0   
3                 TX  OMRON OILFIELD AND MARINE, INC.          65998.0   
4                 VA   FEDERAL HOME LOAN MORTGAGE CO.          96907.0   

     SOC_CODE NAIC_CODE  CASE_DURATION  EMPLOYMENT_DURATION  
0  17-2141.00      3345              0                 19.0  
1     25-1032      6113              6               1095.0  
2     15-1799      5172            711               1094.0  
3     17-2072      3353              6               1096.0  
4     15-1131      5222              6               1095.0

**The data is for EDA. For training, we need further processing on VISA_CLASS (one-hot formation), CASE_NUMBER (delete), Worksite_STATE_abb (one-hot formation), SOC_CODE (one-hot formation), NAIC_CODE (one-hot formation) and EMP_STATE_abb (one-hot formation).**